File: android/app/src/main/java/com/mitimaiti/app/services/SocketManager.kt
package com.mitimaiti.app.services

import io.socket.client.IO
import io.socket.client.Socket
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONObject

/**
 * WebSocket manager for real-time chat.
 * Mirrors: backend/src/socket.ts events
 *
 * Events sent:
 *   send_msg     { matchId, content, msgType }
 *   typing       { matchId }
 *   enter_chat   { matchId }
 *   leave_chat   { matchId }
 *   heartbeat    { }
 *
 * Events received:
 *   new_msg      { id, matchId, senderId, content, msgType, createdAt }
 *   typing       { matchId, userId }
 *   msg_read     { matchId, readBy }
 *   match_update { matchId, status, firstMsgLocked, ... }
 *   new_match    { match object }
 *   error        { message }
 */
class SocketManager private constructor() {
    private var socket: Socket? = null
    private var tokenProvider: (suspend () -> String?)? = null

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    private val _incomingMessages = MutableSharedFlow<JSONObject>(extraBufferCapacity = 64)
    val incomingMessages: SharedFlow<JSONObject> = _incomingMessages.asSharedFlow()

    private val _typingEvents = MutableSharedFlow<JSONObject>(extraBufferCapacity = 64)
    val typingEvents: SharedFlow<JSONObject> = _typingEvents.asSharedFlow()

    private val _readReceipts = MutableSharedFlow<JSONObject>(extraBufferCapacity = 64)
    val readReceipts: SharedFlow<JSONObject> = _readReceipts.asSharedFlow()

    private val _matchUpdates = MutableSharedFlow<JSONObject>(extraBufferCapacity = 16)
    val matchUpdates: SharedFlow<JSONObject> = _matchUpdates.asSharedFlow()

    private val _newMatches = MutableSharedFlow<JSONObject>(extraBufferCapacity = 16)
    val newMatches: SharedFlow<JSONObject> = _newMatches.asSharedFlow()

    fun init(getToken: suspend () -> String?) {
        this.tokenProvider = getToken
    }

    fun connect(token: String) {
        if (socket?.connected() == true) return

        val opts = IO.Options().apply {
            auth = mapOf("token" to token)
            transports = arrayOf("websocket", "polling")
            reconnection = true
            reconnectionAttempts = 10
            reconnectionDelay = 1000
            reconnectionDelayMax = 5000
            timeout = 20000
        }

        socket = IO.socket(ApiConfig.SOCKET_URL, opts).apply {
            on(Socket.EVENT_CONNECT) {
                _isConnected.value = true
                emit("heartbeat", JSONObject())
            }

            on(Socket.EVENT_DISCONNECT) {
                _isConnected.value = false
            }

            on("new_msg") { args ->
                if (args.isNotEmpty()) {
                    _incomingMessages.tryEmit(args[0] as JSONObject)
                }
            }

            on("typing") { args ->
                if (args.isNotEmpty()) {
                    _typingEvents.tryEmit(args[0] as JSONObject)
                }
            }

            on("msg_read") { args ->
                if (args.isNotEmpty()) {
                    _readReceipts.tryEmit(args[0] as JSONObject)
                }
            }

            on("match_update") { args ->
                if (args.isNotEmpty()) {
                    _matchUpdates.tryEmit(args[0] as JSONObject)
                }
            }

            on("new_match") { args ->
                if (args.isNotEmpty()) {
                    _newMatches.tryEmit(args[0] as JSONObject)
                }
            }

            on("error") { args ->
                // Log socket errors
                android.util.Log.e("SocketManager", "Socket error: ${args.firstOrNull()}")
            }

            connect()
        }
    }

    fun disconnect() {
        socket?.disconnect()
        socket = null
        _isConnected.value = false
    }

    // ── Send events ──

    fun sendMessage(matchId: String, content: String, msgType: String = "text") {
        socket?.emit("send_msg", JSONObject().apply {
            put("matchId", matchId)
            put("content", content)
            put("msgType", msgType)
        })
    }

    fun sendTyping(matchId: String) {
        socket?.emit("typing", JSONObject().apply {
            put("matchId", matchId)
        })
    }

    fun enterChat(matchId: String) {
        socket?.emit("enter_chat", JSONObject().apply {
            put("matchId", matchId)
        })
    }

    fun leaveChat(matchId: String) {
        socket?.emit("leave_chat", JSONObject().apply {
            put("matchId", matchId)
        })
    }

    fun sendHeartbeat() {
        socket?.emit("heartbeat", JSONObject())
    }

    companion object {
        val shared = SocketManager()
    }
}

File: ios/MitiMaiti/Services/SocketManager.swift
import Foundation
import SocketIO

/// WebSocket manager for real-time chat. Mirrors backend/src/socket.ts events.
///
/// Events sent: send_msg, typing, enter_chat, leave_chat, heartbeat
/// Events received: new_msg, typing, msg_read, match_update, new_match, error
@MainActor
final class SocketChat: ObservableObject {
    static let shared = SocketChat()

    @Published private(set) var isConnected = false

    private var manager: SocketManager?
    private var socket: SocketIOClient?

    private(set) var incomingMessages = AsyncStream<[String: Any]>.makeStream()
    private(set) var typingEvents = AsyncStream<[String: Any]>.makeStream()
    private(set) var readReceipts = AsyncStream<[String: Any]>.makeStream()
    private(set) var matchUpdates = AsyncStream<[String: Any]>.makeStream()
    private(set) var newMatches = AsyncStream<[String: Any]>.makeStream()

    func connect(token: String) {
        if socket?.status == .connected { return }
        guard let url = URL(string: AppConfig.socketURL) else { return }

        manager = SocketManager(socketURL: url, config: [
            .log(false),
            .compress,
            .reconnects(true),
            .reconnectAttempts(10),
            .reconnectWait(1),
            .reconnectWaitMax(5),
            .connectParams(["token": token])
        ])
        socket = manager?.defaultSocket

        socket?.on(clientEvent: .connect) { [weak self] _, _ in
            self?.isConnected = true
            self?.socket?.emit("heartbeat")
        }
        socket?.on(clientEvent: .disconnect) { [weak self] _, _ in
            self?.isConnected = false
        }
        socket?.on("new_msg") { [weak self] data, _ in
            if let payload = data.first as? [String: Any] {
                self?.incomingMessages.continuation.yield(payload)
            }
        }
        socket?.on("typing") { [weak self] data, _ in
            if let payload = data.first as? [String: Any] {
                self?.typingEvents.continuation.yield(payload)
            }
        }
        socket?.on("msg_read") { [weak self] data, _ in
            if let payload = data.first as? [String: Any] {
                self?.readReceipts.continuation.yield(payload)
            }
        }
        socket?.on("match_update") { [weak self] data, _ in
            if let payload = data.first as? [String: Any] {
                self?.matchUpdates.continuation.yield(payload)
            }
        }
        socket?.on("new_match") { [weak self] data, _ in
            if let payload = data.first as? [String: Any] {
                self?.newMatches.continuation.yield(payload)
            }
        }
        socket?.on("error") { data, _ in
            print("[SocketChat] error:", data)
        }

        socket?.connect()
    }

    func disconnect() {
        socket?.disconnect()
        socket = nil
        manager = nil
        isConnected = false
    }

    // MARK: - Send events

    func sendMessage(matchId: String, content: String, msgType: String = "text") {
        socket?.emit("send_msg", ["matchId": matchId, "content": content, "msgType": msgType])
    }

    func sendTyping(matchId: String) {
        socket?.emit("typing", ["matchId": matchId])
    }

    func enterChat(matchId: String) {
        socket?.emit("enter_chat", ["matchId": matchId])
    }

    func leaveChat(matchId: String) {
        socket?.emit("leave_chat", ["matchId": matchId])
    }

    func sendHeartbeat() {
        socket?.emit("heartbeat")
    }
}
